File: core/src/main/java/no/elg/infiniteBootleg/MouseLocator.java
package no.elg.infiniteBootleg;

import static no.elg.infiniteBootleg.world.Block.BLOCK_SIZE;

import com.badlogic.gdx.Gdx;
import com.badlogic.gdx.math.Vector2;
import com.badlogic.gdx.math.Vector3;
import no.elg.infiniteBootleg.util.CoordUtil;
import no.elg.infiniteBootleg.world.ClientWorld;
import no.elg.infiniteBootleg.world.box2d.WorldBody;
import org.jetbrains.annotations.NotNull;

public class MouseLocator {

  private int mouseBlockX;
  private int mouseBlockY;
  private float mouseWorldX;
  private float mouseWorldY;

  private int previousMouseBlockX;
  private int previousMouseBlockY;
  private float previousMouseWorldX;
  private float previousMouseWorldY;

  private final Vector2 mouseWorldInput = new Vector2();
  private final Vector3 screenInputVec = new Vector3();

  void update(ClientWorld world) {
    screenInputVec.set(Gdx.input.getX(), Gdx.input.getY(), 0);
    world.getRender().getCamera().unproject(screenInputVec);
    // Whenever z is not zero unproject returns a very low number
    // I don't know why this is the case, but checking for z to be zero seems to fix the bug
    if (screenInputVec.z == 0f) {

      previousMouseWorldX = mouseWorldX;
      previousMouseWorldY = mouseWorldY;
      previousMouseBlockX = mouseBlockX;
      previousMouseBlockY = mouseBlockY;

      WorldBody worldBody = world.getWorldBody();
      mouseWorldX = screenInputVec.x / BLOCK_SIZE - worldBody.getWorldOffsetX();
      mouseWorldY = screenInputVec.y / BLOCK_SIZE - worldBody.getWorldOffsetY();
      mouseWorldInput.set(mouseWorldX, mouseWorldY);

      mouseBlockX = CoordUtil.worldToBlock(mouseWorldX);
      mouseBlockY = CoordUtil.worldToBlock(mouseWorldY);
    }
  }

  public int getMouseBlockX() {
    return mouseBlockX;
  }

  public int getMouseBlockY() {
    return mouseBlockY;
  }

  public float getMouseWorldX() {
    return mouseWorldX;
  }

  public float getMouseWorldY() {
    return mouseWorldY;
  }

  public int getPreviousMouseBlockX() {
    return previousMouseBlockX;
  }

  public int getPreviousMouseBlockY() {
    return previousMouseBlockY;
  }

  public float getPreviousMouseWorldX() {
    return previousMouseWorldX;
  }

  public float getPreviousMouseWorldY() {
    return previousMouseWorldY;
  }

  @NotNull
  public Vector2 getMouse() {
    return mouseWorldInput;
  }
}

File: core/src/main/kotlin/no/elg/infiniteBootleg/input/MouseLocator.kt
package no.elg.infiniteBootleg.input

import com.badlogic.gdx.Gdx
import com.badlogic.gdx.math.Vector2
import com.badlogic.gdx.math.Vector3
import no.elg.infiniteBootleg.util.WorldCoord
import no.elg.infiniteBootleg.util.WorldCoordFloat
import no.elg.infiniteBootleg.util.compactLoc
import no.elg.infiniteBootleg.util.worldToBlock
import no.elg.infiniteBootleg.world.blocks.Block
import no.elg.infiniteBootleg.world.world.ClientWorld

class MouseLocator {
  var mouseBlockX: WorldCoord = 0
    private set
  var mouseBlockY: WorldCoord = 0
    private set
  var mouseWorldX: WorldCoordFloat = 0f
    private set
  var mouseWorldY: WorldCoordFloat = 0f
    private set
  var previousMouseBlockX: WorldCoord = 0
    private set
  var previousMouseBlockY: WorldCoord = 0
    private set
  var previousMouseWorldX: WorldCoordFloat = 0f
    private set
  var previousMouseWorldY: WorldCoordFloat = 0f
    private set

  private val mouseWorldInput = Vector2()
  private val screenInputVec = Vector3()

  val mouseBlockCompactLoc: Long get() = compactLoc(mouseBlockX, mouseBlockY)

  /**
   * If the block the mouse is over has changed since the last update
   */
  val isOverNewBlock: Boolean get() = mouseBlockX != previousMouseBlockX || mouseBlockY != previousMouseBlockY

  fun update(world: ClientWorld) {
    screenInputVec.set(Gdx.input.x.toFloat(), Gdx.input.y.toFloat(), 0f)
    world.render.camera.unproject(screenInputVec)
    // Whenever z is not zero unproject returns a very low number
    // I don't know why this is the case, but checking for z to be zero seems to fix the bug
    if (screenInputVec.z == 0f) {
      previousMouseWorldX = mouseWorldX
      previousMouseWorldY = mouseWorldY
      previousMouseBlockX = mouseBlockX
      previousMouseBlockY = mouseBlockY

      mouseWorldX = screenInputVec.x / Block.BLOCK_SIZE
      mouseWorldY = screenInputVec.y / Block.BLOCK_SIZE
      mouseWorldInput.set(mouseWorldX, mouseWorldY)
      mouseBlockX = worldToBlock(mouseWorldX)
      mouseBlockY = worldToBlock(mouseWorldY)
    }
  }
}
